File: client-clojure/src/sirqul_io_t_platform/specs/application_usage_response.clj
(ns sirqul-io-t-platform.specs.application-usage-response
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [sirqul-io-t-platform.specs.billable-entity-response :refer :all]
            )
  (:import (java.io File)))


(def application-usage-response-data
  {
   (ds/opt :billableEntity) billable-entity-response-spec
   (ds/opt :applicationId) int?
   (ds/opt :invoiceDate) inst?
   (ds/opt :balance) float?
   (ds/opt :totalRequests) int?
   (ds/opt :totalDataStorage) int?
   (ds/opt :totalMediaStorage) int?
   (ds/opt :totalNotifications) int?
   (ds/opt :maxMissions) int?
   (ds/opt :maxVouchers) int?
   (ds/opt :maxCoupons) int?
   (ds/opt :maxProducts) int?
   (ds/opt :maxEvents) int?
   (ds/opt :maxMedia) int?
   (ds/opt :maxLocations) int?
   })

(def application-usage-response-spec
  (ds/spec
    {:name ::application-usage-response
     :spec application-usage-response-data}))

File: client-scalaz/src/main/scala/org/openapitools/client/api/ApplicationUsageResponse.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime

import java.time.OffsetDateTime

import ApplicationUsageResponse._

case class ApplicationUsageResponse (
  billableEntity: Option[BillableEntityResponse],
applicationId: Option[Long],
invoiceDate: Option[OffsetDateTime],
balance: Option[Double],
totalRequests: Option[Long],
totalDataStorage: Option[Long],
totalMediaStorage: Option[Long],
totalNotifications: Option[Long],
maxMissions: Option[Long],
maxVouchers: Option[Long],
maxCoupons: Option[Long],
maxProducts: Option[Long],
maxEvents: Option[Long],
maxMedia: Option[Long],
maxLocations: Option[Long])

object ApplicationUsageResponse {
  import DateTimeCodecs._

  implicit val ApplicationUsageResponseCodecJson: CodecJson[ApplicationUsageResponse] = CodecJson.derive[ApplicationUsageResponse]
  implicit val ApplicationUsageResponseDecoder: EntityDecoder[ApplicationUsageResponse] = jsonOf[ApplicationUsageResponse]
  implicit val ApplicationUsageResponseEncoder: EntityEncoder[ApplicationUsageResponse] = jsonEncoderOf[ApplicationUsageResponse]
}
